src/gui/widgets/FxLine.cpp

```diff
@@ -395,3 +415,57 @@ void FxLine::setStrokeInnerInactive( const QColor & c )
 {
 	m_strokeInnerInactive = c;
 }
+
+void FxLine::changeColor()
+{
+	QColorDialog colorDialog( Engine::fxMixer()->effectChannel( m_channelIndex )->m_color );
+	QColor buffer( 0, 0, 0 );
+	
+	for( int i = 0; i < 48; i += 6 )
+	{
+		for( int j = 0; j < 6; j++ )
+		{
+			buffer.setHsl( qMax( 0, 44 * ( i / 6 ) - 1 ), 150 - 20 * j, 140 - 10 * j );
+			colorDialog.setStandardColor( i + j, buffer );
+		}
+		
+	}
+	
+	QColor new_color = colorDialog.getColor( Engine::fxMixer()->effectChannel( m_channelIndex )->m_color );
+	if( ! new_color.isValid() )
+	{ return; }
+	
+	Engine::fxMixer()->effectChannel( m_channelIndex )->m_color = new_color;
+	Engine::fxMixer()->effectChannel( m_channelIndex )->m_hasColor = true;
+	
+	update();
+}
+
+void FxLine::resetColor()
+{
+	Engine::fxMixer()->effectChannel( m_channelIndex )->m_hasColor = false;
+	update();
+}
+
+void FxLine::randomColor()
+{
+	int index = rand() % 48;
+	QColor buffer( 0, 0, 0 );
+	
+	for( int i = 0; i < 48; i += 6 )
```

Comment:
Considering the colors are algorithmically derived, saving them would just take up memory. I'd say if ColorChooser has a getter method, then I can switch to that (or else just use the color at index function of the color dialog)

Reply:
Ok, we could go with a function that returns the color for a given index i. Rather than creating the whole list and then selecting from it, find x and y from the index and then use them directly in the equation:

```
x = i % 6;
y = i / 6;
return QColor::fromHsv( qMax( 0, 44 * y - 1 ), 150 - 20 * x, 140 - 10 * x );
```

I don't think you should rely on the color-at-index function of the color chooser itself, because then you need to instantiate a color chooser even when you don't need one. That seems significantly heavier than accessing an existing list or using the getter described above.

Sidenote: assuming QColor takes 10x the space of three ints for some reason, 48 QColors is 5760 bytes, which is less than 0.0062% of a gigabyte. That means we could store 173 thousand+ copies of those 48 colors in a gigabyte. I don't think the memory use of storing this list is a concern.

Reply:
I know I'm late, but ColorChooser doesn't even have a .cpp file afaik. I guess I need to make one now...

Reply:
Implemented a workaround, making FxLine store one QColorDialog for common usage, so it's more efficient and I can just grab the index from there. No useless initialisations, it's initialised along with FxLine.

Reply:
Resolving this conversation since it's basically the same as [this one](https://github.com/LMMS/lmms/pull/5589#discussion_r459654083).